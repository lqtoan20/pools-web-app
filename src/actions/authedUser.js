export const SET_AUTHED_USER = "SET_AUTHED_USER";
export const LOGOUT_AUTHED_USER = "LOGOUT_AUTHED_USER";

// Action creators 1
export function setAuthedUser(authedUser) {
  return {
    type: SET_AUTHED_USER,
    authedUser,
  };
}

export function logoutAuthedUser() {
  return {
    type: LOGOUT_AUTHED_USER,
  };
}

// Thunk actions
export function handleLogin(username, password) {
  return (dispatch, getState) => {
    const { users } = getState();

    const user = Object.values(users).find(
      (user) => user.id === username && user.password === password
    );

    if (user) {
      dispatch(setAuthedUser(user));
    } else {
      // Handle login failure here (e.g., show an error message)
      console.log("Login failed: Invalid username or password");
    }
  };
}

export function handleLogout() {
  return (dispatch) => {
    dispatch(logoutAuthedUser());
    // Perform additional logout-related actions if needed
  };
}
