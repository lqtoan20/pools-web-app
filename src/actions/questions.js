import { saveQuestion, saveQuestionAnswer } from "../helper/api";
import { addAnswerUser, addQuestionUser } from "./users";

export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const ADD_QUESTION = "ADD_QUESTION";
export const ADD_ANSWER_QUESTION = "ADD_ANSWER_QUESTION";

// Action creators
export function receiveQuestions(questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

export function addQuestion(question) {
  return {
    type: ADD_QUESTION,
    question,
  };
}

export function addAnswerQuestion(author, qid, answer) {
  return {
    type: ADD_ANSWER_QUESTION,
    author,
    qid,
    answer,
  };
}

// Thunk actions
export function handleAddQuestion(firstOption, secondOption) {
  return (dispatch, getState) => {
    const { authedUser } = getState();

    return saveQuestion(firstOption, secondOption, authedUser)
      .then((question) => {
        dispatch(addQuestion(question));
        dispatch(addQuestionUser(question));
      })
      .catch((error) => {
        // Handle error here
        console.error("Error adding question:", error);
      });
  };
}

export function handleAddAnswer(questionId, answer) {
  return (dispatch, getState) => {
    const { authedUser } = getState();

    return saveQuestionAnswer(authedUser.id, questionId, answer)
      .then(() => {
        dispatch(addAnswerQuestion(authedUser.id, questionId, answer));
        dispatch(addAnswerUser(authedUser.id, questionId, answer));
      })
      .catch((error) => {
        // Handle error here
        console.error("Error adding answer:", error);
      });
  };
}
