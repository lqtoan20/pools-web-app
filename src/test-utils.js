import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { store } from "./stores/store";

function Wrapper(children) {
  return render(
    <Provider store={store}>
      <BrowserRouter>{children}</BrowserRouter>
    </Provider>
  );
}
export { Wrapper };
