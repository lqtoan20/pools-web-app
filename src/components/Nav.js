import { Link } from "react-router-dom";
import { connect } from "react-redux";
import { handleLogout } from "../actions/authedUser";

const Nav = ({ dispatch, authedUserId }) => {
  const logout = (e) => {
    e.preventDefault();
    dispatch(handleLogout());
  };

  return (
    <nav className="uk-flex uk-flex-center uk-margin-medium-top">
      <ul className="uk-subnav uk-subnav-divider">
        <li>
          <Link to="/" className="uk-text-bold uk-link-text">
            Home
          </Link>
        </li>
        <li>
          <Link to="/leaderboards" className="uk-text-bold uk-link-text">
            Leaderboard
          </Link>
        </li>
        <li>
          <Link to="/new" className="uk-text-bold uk-link-text">
            New Poll
          </Link>
        </li>
        <li>
          <span
            data-testid="information-data"
            className="uk-text-bold"
          >{`User: ${authedUserId}`}</span>
        </li>
        <li>
          <button
            onClick={logout}
            className="uk-button uk-button-default uk-margin-small-left"
          >
            Logout
          </button>
        </li>
      </ul>
    </nav>
  );
};

const mapStateToProps = ({ authedUser }) => ({
  authedUserId: authedUser.id,
});

export default connect(mapStateToProps)(Nav);
