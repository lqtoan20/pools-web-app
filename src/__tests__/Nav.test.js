import { store } from "../stores/store";
import React from "react";
import Nav from "../components/Nav";
import { setAuthedUser } from "../actions/authedUser";
import { Wrapper } from "../test-utils";

describe("Nav", () => {
  it("should render the component", () => {
    store.dispatch(setAuthedUser({ id: "sarahedo", password: "" }));
    const component = Wrapper(<Nav />);
    expect(component).toBeDefined();
    expect(component).toMatchSnapshot();
  });

  it("should display username of logged in user", () => {
    store.dispatch(setAuthedUser({ id: "sarahedo", password: "" }));

    const component = Wrapper(<Nav />);

    expect(component.getByTestId("information-data").textContent).toBe(
      "User: sarahedo"
    );
  });
});
