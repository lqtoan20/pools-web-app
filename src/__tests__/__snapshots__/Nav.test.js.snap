// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="uk-flex uk-flex-center uk-margin-medium-top"
      >
        <ul
          class="uk-subnav uk-subnav-divider"
        >
          <li>
            <a
              class="uk-text-bold uk-link-text"
              href="/"
            >
              Home
            </a>
          </li>
          <li>
            <a
              class="uk-text-bold uk-link-text"
              href="/leaderboards"
            >
              Leaderboard
            </a>
          </li>
          <li>
            <a
              class="uk-text-bold uk-link-text"
              href="/new"
            >
              New Poll
            </a>
          </li>
          <li>
            <span
              class="uk-text-bold"
              data-testid="information-data"
            >
              User: sarahedo
            </span>
          </li>
          <li>
            <button
              class="uk-button uk-button-default uk-margin-small-left"
            >
              Logout
            </button>
          </li>
        </ul>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="uk-flex uk-flex-center uk-margin-medium-top"
    >
      <ul
        class="uk-subnav uk-subnav-divider"
      >
        <li>
          <a
            class="uk-text-bold uk-link-text"
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            class="uk-text-bold uk-link-text"
            href="/leaderboards"
          >
            Leaderboard
          </a>
        </li>
        <li>
          <a
            class="uk-text-bold uk-link-text"
            href="/new"
          >
            New Poll
          </a>
        </li>
        <li>
          <span
            class="uk-text-bold"
            data-testid="information-data"
          >
            User: sarahedo
          </span>
        </li>
        <li>
          <button
            class="uk-button uk-button-default uk-margin-small-left"
          >
            Logout
          </button>
        </li>
      </ul>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
